mite_selenium/__init__.py

```diff
@@ -51,32 +56,119 @@ def start(self):
     def stop(self):
         self._context.browser.close()
 
-
-def send_selenium_times(context):
-    """Maybe would probably need more error handling
-    if somebody tried to use with an incompatible browser"""
-    if context.browser.capabilities['browserName'] != 'chrome':
-        logger.warning(
-            'Timing metrics in %s are not currently supported',
-            context.browser.capabilities['browserName'],
-        )
-        return
-
-    timings = context.browser.execute_script(
-        'return performance.getEntriesByType("navigation")'
-    )[0]
-    context.send(
-        "http_selenium_metrics",
-        dom_complete=timings["domComplete"],
-        dom_interactive=timings["domInteractive"],
-        transfer_size=timings["transferSize"],
-        transfer_start_timings=["requestStart"],
-        tls_time=timings["secureConnectionStart"],
-        dns_time=timings["domainLookupEnd"],
-        connect_time=timings["connectEnd"],
-        total_time=timings["duration"],
-        effective_url=timings["name"],
-    )
+    def _browser_has_timing_capabilities(self):
+        return self._context.browser.capabilities["browserName"] != "chrome"
+
+    def _is_using_tls(self, name):
+        return name.startsWith("https")
+
+    def _get_tls_timing(self, timings):
+        if self._is_using_tls(timings["name"]):
+            return timings["connectEnd"] - timings["secureConnectionStart"]
+        else:
+            logger.warning("Secure TLS connection not used, defaulting tls_time to 0")
+            return 0
+
+    def _get_tcp_timing(self, timings):
+        if self._is_using_tls(timings["name"]):
+            return timings["secureConnectionStart"] - timings["connectStart"]
+        else:
+            return timings["connectEnd"] - timings["connectStart"]
+
+    def _send_page_load_metrics(self):
+        if self._browser_has_timing_capabilities():
+            timings, *_ = self._context.browser.execute_script(
+                "return performance.getEntriesByType('navigation')"
+            )
```

Comment:
I'm pretty sure that this should just be a one-element list.  But we might want to check that and warn if there are more elements.  I'm not totally sure the circumstances under which the perf event buffer is reset, but in principle it could happen that we do `get(foo.com); get(bar.com)` and then the list of navigation perf events is `[foo, bar]` and we mistakenly report the foo stats for bar.

At least until we nail down how this all works, I'd prefer to be defensive here.

Reply:
I'm not quite sure how you mean? In this instance from my understanding, we're taking the head element as timings and ignoring the rest. Is the suggestion to change this to use the following:

`timings = self._context.browser.execute_script(...)[0]`

Reply:
`performance.getEntriesByType('navigation')` returns a list of page navigation timings.  We are assuming that the zeroth element of that list is relevant to the current page.  That's probably correct if we load the page by (a simulated version of) typing the address into the address bar and hitting enter.  There are various JS apis to navigate though that a page could call.  I'm not sure whether all of those would also clear the list, or whether the list would survive across those.  So we might get into a situation where the list returned by `performance.getEntriesByType('navigation')` has multiple entries.  (There's then a subquestion about whether new entries are added to the beginning or the end of the list, but I assume it's add-to-end).

So, we have a packet of assumptions about this list being exactly one element long; if those assumptions are violated we'll get an error -- either a runtime error in python if there are zero elements, or a logic error (recording the wrong timings) if there are more.  I'm saying we should have an explicit test of the returned list length, rather than just assuming it will always be one, and we should log a specific error if that's not fulfilled.

Reply:
I've wrapped this to make sure that we're explicitly only acting on one list item